File: 763. Partition Labels.java
// Time Complexity: O(n)
// Space Complexity: O(1)
// Did this code successfully run on Leetcode: YES
// Any problem you faced while coding this: NO

class Solution {
    public List<Integer> partitionLabels(String s) {
        if(s == null || s.length() == 0) return new ArrayList<>();
        
        HashMap<Character, Integer> map = new HashMap<>();
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            map.put(c, i);
        }
        List<Integer> result = new ArrayList<>();
        
        int start = 0, end = 0;
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            end = Math.max(end, map.get(c));
            if(i == end){
                result.add(end - start + 1);
                start = end + 1;
            }
        }
        return result;
    }
}

File: 763-PartitionLabels.cpp
/* 
    Time Complexity                              :  O(2*N) used in the 2 for loops. 
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    vector<int> partitionLabels(string s) {
        int n = s.size();
        vector<int> ans;
        vector<int> chars(26,0);
        for(int i=0;i<n;i++) {
            int idx = s[i] - 'a';
            chars[idx] = i;
        }
        
        int mx=0,count=0;
        for(int i=0;i<n;i++) {
            int idx = s[i]-'a';
            int cmx = chars[idx];
            
            
            if(cmx > mx) {
                mx = cmx;
            } 
            
            count++;
            if(i == mx) {
                ans.push_back(count);
                count=0;
            }
        }
        
        return ans;
    }
};